### DISCUSSION TITLE
/api/queries CORS error while working on localhost

### DISCUSSION THREAD

**Discussion by andresMRoblesistemas** [getredash/redash#7182] [Q&A]
Im working in localhost to test redash.
When I try it in Postman it works. The queries are created on the instance of redash

![image](https://github.com/user-attachments/assets/cfed196d-f5b5-4da7-8fcf-3b0f516b9b81)
I'm creating a redash query and then when I want to send it to /api/queries I use the following method (Dart code)

```dart
Future<String> getRedashResponse() async {
        // URL de la API de Redash
        final url = Uri.http(
          'localhost:5001',
          '/api/queries',
        );
        final body = jsonEncode({
          "query": "{\n    \"collection\":\"cliente\"\n}",
          "name": "charla ${DateTime.now()}",
          "data_source_id": 1,
          "options": {"parameters": []}
        });
        final headers = {
          'Content-Type': 'application/json',
          'Authorization': 'Key fgdffddfgdfgf',
          'Access-Control-Allow-Origin': 'strict-origin-when-cross-origin'
        };

        final response = await http
            .post(url, headers: headers, body: body)
            .timeout(const Duration(minutes: 1));

        if (response.statusCode == 200) {
          final responseJson = jsonDecode(response.body);
          final jsonString = responseJson['data'][0]['respuesta'];
          return jsonString;
        } else {
          log('Request failed with status: ${response.statusCode}');
          return 'Error: ${response.statusCode}';
        }
      }
```
But I get the error in the image above. How do I fix that?

**Comment by justinclift** [comment]
Which version of Redash are you using, and what process did you use to set it up? :smile:

Asking because for local development purposes (ie to do Redash development) there are guides here which are known to work: https://github.com/getredash/redash/wiki

For a standalone system that you're preparing in a production like way (even just testing in a VM), then the setup.sh script here should be good too: https://github.com/getredash/setup/

**Reply by andresMRoblesistemas** [reply]
I'm using the Ubuntu guide from the wiki. https://github.com/getredash/redash/wiki/Local-development-setup
So I guess I'm using the last version. Redash web_ui is working, I can save queries and make viz. From Postman I can use the api too. But making the POST from my dart code doesn't work. It give the CORS error

**Reply by justinclift** [reply]
Interesting.  I wonder what Dart is doing differently?

You *might* need to do a packet capture of the complete request via Dart, and via Postman, to find out what the difference is.

That being said, I wonder if the `REDASH_ENFORCE_CSRF` environment variable (or one of the others) might be mucking things up for you?

* https://redash.io/help/open-source/admin-guide/env-vars-settings/#REDASH_ENFORCE_CSRF

**Reply by justinclift** [reply]
There are some CORS related environment variables on that page as well, so it might be useful to experiment with those too. :smile:

**Reply by andresMRoblesistemas** [reply]
I've added a nginx.conf file and the environment variables too in the compose.yaml
[docker-compose.yaml.txt](https://github.com/user-attachments/files/17308732/docker-compose.yaml.txt)
[nginx.conf.txt](https://github.com/user-attachments/files/17308733/nginx.conf.txt)

**Reply by andresMRoblesistemas** [reply]
Still having the same issue
![image](https://github.com/user-attachments/assets/9510c8ff-6058-437b-9c3d-aa78201e52c4)
![image](https://github.com/user-attachments/assets/fbe12c3c-52af-4a22-b26d-ec23777fdd06)

**Reply by justinclift** [reply]
Hmmm, the error mentions 'Access-Control-Allow-Origin'.  There's a Redash environment variable for that too (`REDASH_CORS_ACCESS_CONTROL_ALLOW_ORIGIN`), though it doesn't look to be one of the ones you've added to your docker-compose.yaml file?

**Reply by andresMRoblesistemas** [reply]
I've tried it and it says the same error (Ive deleted every image, container and all, Ive only replace the original compose.yaml for the one with all the environment variables) Im giving up, Im moving forward with Metabase instead.

**Reply by justinclift** [reply]
No worries.  It would have been interesting for you to capture the full request as made by Dart and by something that does it properly (ie you mention Postman), to see what the difference is and what Dart's doing wrong.

Hopefully Metabase works for you out of the box instead. :smile:
